ipie/addons/thermal/walkers/pop_controller.py

```diff
@@ -0,0 +1,442 @@
+
+from ipie.config import MPI
+from ipie.utils.backend import arraylib as xp
+from ipie.walkers.pop_controller import PopController, PopControllerTimer
+from ipie.addons.thermal.walkers.uhf_walkers import UHFThermalWalkers
+
+
+class ThermalPopController(PopController):
+    def __init__(
+        self,
+        num_walkers_local,
+        num_steps,
+        mpi_handler=None,
+        pop_control_method="pair_branch",
+        min_weight=0.1,
+        max_weight=4,
+        reconfiguration_freq=50,
+        verbose=False,
+    ):
+        super().__init__(
+                num_walkers_local, num_steps, mpi_handler, pop_control_method,
+                min_weight, max_weight, reconfiguration_freq, verbose)
+
+    def pop_control(self, walkers, comm):
+        self.timer.start_time()
+        if self.ntot_walkers == 1:
+            return
+        weights = numpy.abs(xp.array(walkers.weight))
+        global_weights = numpy.empty(len(weights) * comm.size)
+        self.timer.add_non_communication()
+        self.timer.start_time()
+        if self.method == "comb":
+            comm.Allgather(weights, global_weights)
+            total_weight = sum(global_weights)
+        else:
+            sum_weights = numpy.sum(weights)
+            total_weight = numpy.empty(1, dtype=numpy.float64)
+            if hasattr(sum_weights, "get"):
+                sum_weights = sum_weights.get()
+            comm.Reduce(sum_weights, total_weight, op=MPI.SUM, root=0)
+            comm.Bcast(total_weight, root=0)
+            total_weight = total_weight[0]
+
+        self.timer.add_communication()
+        self.timer.start_time()
+
+        # Rescale weights to combat exponential decay/growth.
+        scale = total_weight / self.target_weight
+        if total_weight < 1e-8:
+            if comm.rank == 0:
+                print(f"# Warning: Total weight is {total_weight:13.8e}")
+                print("# Something is seriously wrong.")
+            raise ValueError
+        self.total_weight = total_weight
+        # Todo: Just standardise information we want to send between routines.
+        walkers.unscaled_weight = walkers.weight
+        walkers.weight = walkers.weight / scale
+        if self.method == "comb":
+            global_weights = global_weights / scale
+            self.timer.add_non_communication()
+            comb(walkers, comm, global_weights, self.target_weight, self.timer)
+        elif self.method == "pair_branch":
+            pair_branch(walkers, comm, self.max_weight, self.min_weight, self.timer)
+        elif self.method == "stochastic_reconfiguration":
+            self.reconfiguration_counter += 1
+            if self.reconfiguration_counter % self.reconfiguration_freq == 0:
+                stochastic_reconfiguration(walkers, comm, self.timer)
+                self.reconfiguration_counter = 0
+        else:
+            if comm.rank == 0:
+                print("Unknown population control method.")
+
+
+def get_buffer(walkers, iw):
```

Comment:
Why is any of this code necessary? The population control algorithms should be agnostic to the underlying qmc algorithm. If it's not the case then that's a problem and we should open an issue. But the get/set buffers which are the only potentially algorithm dependent quantities should be delegated to the walkers right?

Reply:
Right, the walker should do the concatenation. Can you open an issue to track this.